server/mailet/rate-limiter/src/main/scala/org/apache/james/transport/mailets/PerRecipientRateLimitMailet.scala

```diff
@@ -0,0 +1,104 @@
+/****************************************************************
+ * Licensed to the Apache Software Foundation (ASF) under one   *
+ * or more contributor license agreements.  See the NOTICE file *
+ * distributed with this work for additional information        *
+ * regarding copyright ownership.  The ASF licenses this file   *
+ * to you under the Apache License, Version 2.0 (the            *
+ * "License"); you may not use this file except in compliance   *
+ * with the License.  You may obtain a copy of the License at   *
+ *                                                              *
+ * http://www.apache.org/licenses/LICENSE-2.0                   *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.transport.mailets
+
+import java.time.Duration
+import java.time.temporal.ChronoUnit
+
+import com.google.common.annotations.VisibleForTesting
+import javax.inject.{Inject, Named}
+import org.apache.james.rate.limiter.api.{AcceptableRate, AllowedQuantity, RateExceeded, RateLimiter, RateLimiterFactory, RateLimiterFactoryProvider, RateLimitingKey, RateLimitingResult, Rule, Rules}
+import org.apache.mailet.Mail
+import org.apache.mailet.base.GenericMailet
+import org.apache.james.core.MailAddress
+import org.apache.james.util.DurationParser
+import org.reactivestreams.Publisher
+import reactor.core.scala.publisher.{SFlux, SMono}
+
+case class PerRecipientRateLimiter(rateLimiter: RateLimiter, keyPrefix: Option[KeyPrefix], entityType: EntityType) {
+  def rateLimit(recipient: MailAddress, mail: Mail): Publisher[RateLimitingResult] =
+    entityType.extractQuantity(mail)
+      .map(increment => rateLimiter.rateLimit(RecipientKey(keyPrefix, entityType, recipient), increment))
+      .getOrElse(SMono.just[RateLimitingResult](RateExceeded))
+}
+
+case class RecipientKey(keyPrefix: Option[KeyPrefix], entityType: EntityType, mailAddress: MailAddress) extends RateLimitingKey {
+  override def asString(): String = s"${
+    keyPrefix.map(prefix => prefix.value + "_")
+      .getOrElse("")
+  }${entityType.asString()}_${mailAddress.asString()}"
+}
+
+class PerRecipientRateLimitMailet @Inject()(@Named("perRecipientRateLimiterFactoryProvider") rateLimiterFactoryProvider: RateLimiterFactoryProvider) extends GenericMailet {
```

Comment:
IMO we do not need named injects: all instances creating rateLimiters could use the very same factory.